File: leetcode/python/0290_word_pattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        # a -> dog, a -> fish NO
        # a -> dog, b -> dog NO
        # i to traverse through pattern, word to contain substring of s between start and j
        # letterMap
        # traverse through s using j:
        # if s[j] == " " or j == len(s) - 1:
        #   if s[j] == " ": word = s[start:j]
        #   elif j == len(s) - 1: word = s[start:]
        #   if pattern[i] in letterMap.keys() and pattern[i] != word: return False
        #   letterMap[pattern[i]] = word
        #   i += 1, start = j + 1
        # return True
        letterMap, i, start = {}, 0, 0
        for j in range(0, len(s)):
            if i > len(pattern) - 1: return False

            if s[j] == " " or j == len(s) - 1:
                if s[j] == " ": word = s[start:j]
                elif j == len(s) - 1: word = s[start:]
                if pattern[i] in letterMap.keys() and letterMap[pattern[i]] != word: return False
                if word in letterMap.values() and not pattern[i] in letterMap.keys(): return False 
                letterMap[pattern[i]] = word
                i += 1
                start = j + 1

        if i != len(pattern): return False

        return True

File: leetcode/java/0290_word_pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        // patternMap can only cotain:
        // a -> dog, a -> cat x
        // a -> dog, b -> dog yes
        
        // patternMap contains patterns between pattern[i] and sArray[i]
        // look through pattern using i:
        // if pattern[i] is not in hashPattern: add it
        // if pattern[i] is in hashPattern:
        // -> if hashPattern.get(pattern[i]) != sArray[i] return false
        // return true

        Map<Character, String> hashPattern = new HashMap<Character, String>();
        String[] sArray = s.split("\\s+");

        if (sArray.length != pattern.length()) return false;

        for (int i = 0; i < pattern.length(); i++) {
            if (hashPattern.containsKey(pattern.charAt(i)) == false) {
                if (hashPattern.containsValue(sArray[i]) == true) return false;
                hashPattern.put(pattern.charAt(i), sArray[i]);
            } else if (hashPattern.containsKey(pattern.charAt(i)) == true) {
                if (!hashPattern.get(pattern.charAt(i)).equals(sArray[i])) {
                    return false;
                }
            }
        }

        return true;
    }
}